src/go.ts

```diff
@@ -0,0 +1,298 @@
+
+  /**
+   * The username to use for the commit.
+   *
+   * @default - taken from git config. throws if not configured.
+   */
+  readonly username?: string;
+
+  /**
+   * The email to use for the commit.
+   *
+   * @default - taken from git config. throws if not configured.
+   */
+  readonly email?: string;
+
+  /**
+   * The version.
+   *
+   * @default - taken from the 'version'. throws if doesn't exist.
+   */
+  readonly version?: string;
+}
+
+/**
+ * Information about a specific module.
+ */
+export interface GoModule {
+
+  /**
+   * Path to the mod file.
+   */
+  readonly modFile: string;
+
+  /**
+   * The version of the module.
+   */
+  readonly version: string;
+
+  /**
+   * The cannonical name of the module. (e.g 'github.com/aws/constructs-go/constructs/v3`)
+   */
+  readonly cannonicalName: string;
+
+  /**
+   * The path inside the repository the module is located in. (e.g 'constructs')
+   */
+  readonly repoPath: string;
+
+  /**
+   * The repository URL. (e.g 'github.com/aws/constructs')
+   */
+  readonly repoURL: string;
+
+}
+
+/**
+ * Release a set of Golang modules.
+ */
+export class GoReleaser {
+
+  private readonly version?: string;
+
+  private readonly dir: string;
+  private readonly dryRun: boolean;
+  private readonly gitBranch: string;
+  private readonly gitUsername: string;
+  private readonly gitUseremail: string;
+
+  constructor(props: GoReleaserProps) {
+
+    try {
+      utils.checkProgram('git');
+    } catch (err) {
+      throw new Error(`git must be available to create this release: ${err.message}`);
+    }
+
+    this.version = props.version;
+    this.dir = props.dir ?? path.join(process.cwd(), 'dist/go');
+    this.gitBranch = props.branch ?? 'main';
+    this.dryRun = props.dryRun ?? false;
+    this.gitUsername = props.username ?? utils.shell('git config user.name', { capture: true });
+    this.gitUseremail = props.email ?? utils.shell('git config user.email', { capture: true });
+
+    if (this.gitUseremail === '') {
+      throw new Error('Unable to detect username. either configure a global git user.name or pass GIT_USER_NAME env variable');
+    }
+
+    if (this.gitUsername === '') {
+      throw new Error('Unable to detect user email. either configure a global git user.email or pass GIT_USER_EMAIL env variable');
+    }
+
+  }
+
+  /**
+   * Run the release process.
+   *
+   * @returns metadata about the release.
+   */
+  public release(): GoRelease {
+
+    const modules = this.collectModules(this.dir);
+
+    if (modules.length === 0) {
+      console.log('No modules detected. Skipping');
+      return {};
+    }
+
+    console.log('Detected modules:');
+    modules.forEach(m => console.log(` - ${m.modFile}`));
+
+    const repoURL = this.extractRepoURL(modules);
+    const repoDir = path.join(utils.makeTempDirectory(), 'repo');
+    utils.gitHubClone(repoURL, repoDir);
+
+    process.chdir(repoDir);
+
+    try {
+      utils.shell(`git checkout ${this.gitBranch}`);
```

Comment:
I believe -B should do the trick, no?

Reply:
Yeah, I tried and had some issues, but can't repro now. I'll switch.